include/aspect/particle/world.h

```diff
@@ -21,729 +21,310 @@
          *
-         * @param [in] solution The current solution vector for this
-         * simulation.
+         * @param [in] send_particles All particles that should be send and
+         * their new subdomain_ids are in this map.
          */
-        void get_particle_velocities(const LinearAlgebra::BlockVector &solution)
-        {
-          Vector<double>                single_res(dim+2);
-          std::vector<Vector<double> >  result;
-          Point<dim>                    velocity;
-          unsigned int                  i, num_cell_particles;
-          LevelInd                      cur_cell;
-          typename std::multimap<LevelInd, T>::iterator  it, sit;
-          typename DoFHandler<dim>::active_cell_iterator  found_cell;
-          std::vector<Point<dim> >      particle_points;
-
-          // Prepare the field function
-          Functions::FEFieldFunction<dim, DoFHandler<dim>, LinearAlgebra::BlockVector> fe_value(*dof_handler, solution, *mapping);
-
-          // Get the velocity for each cell at a time so we can take advantage of knowing the active cell
-          for (it=particles.begin(); it!=particles.end();)
-            {
-              // Save a pointer to the first particle in this cell
-              sit = it;
-
-              // Get the current cell
-              cur_cell = it->first;
-
-              // Resize the vectors to the number of particles in this cell
-              num_cell_particles = particles.count(cur_cell);
-              particle_points.resize(num_cell_particles);
-
-              // Get a vector of the particle locations in this cell
-              i=0;
-              while (it != particles.end() && it->first == cur_cell)
-                {
-                  if (it->second.vel_check()) particle_points[i++] = it->second.get_location();
-                  it++;
-                }
-              result.resize(i, single_res);
-              particle_points.resize(i);
-
-              // Get the cell the particle is in
-              found_cell = typename DoFHandler<dim>::active_cell_iterator(triangulation, cur_cell.first, cur_cell.second, dof_handler);
-
-              // Interpolate the velocity field for each of the particles
-              fe_value.set_active_cell(found_cell);
-              fe_value.vector_value_list(particle_points, result);
-
-              // Copy the resulting velocities to the appropriate vector
-              it = sit;
-              i = 0;
-              while (it != particles.end() && it->first == cur_cell)
-                {
-                  if (it->second.vel_check())
-                    {
-                      for (int d=0; d<dim; ++d) velocity(d) = result[i](d);
-                      it->second.set_velocity(velocity);
-                      i++;
-                    }
-                  it++;
-                }
-            }
-        };
+        void
+        send_recv_particles(const std::multimap<types::subdomain_id,Particle <dim> > &send_particles);
```

Comment:
They are immediately inserted in the member variable `particles` where they belong. I updated the documentation.